windows/src/developer/TIKE/compile/CompileKeymanWeb.pas

```diff
@@ -1641,19 +1655,17 @@ function TCompileKeymanWeb.ValidateLayoutFile(var sLayoutFile: string; const sVK
       //
       if FText.StartsWith('*') and FText.EndsWith('*') and (FText.Length > 2) then
       begin
+        // We do not support '*special*' labels on non-special keys in Keyman 13
+        // or earlier, except for ZWNJ, which will be transformed in function
+        // TransformSpecialKeys14 to '<|>', and  which does not require the custom
+        // OSK font.
         if (CSpecialText10.Contains(FText) or CSpecialText14.Contains(FText)) and
+            not CSpecialText14ZWNJ.Contains(FText) and
             not IsKeyboardVersion14OrLater and
             not (FKeyType in [tktSpecial, tktSpecialActive]) then
         begin
           ReportError(0, CWARN_TouchLayoutSpecialLabelOnNormalKey,
-            Format('Key "%s" on layout "%s", platform "%s" is not a Special key but has the special label "%s". This feature is only supported in Keyman 14 or later', [
-              FId, FLayer.Id, FPlatform.Name, FText
-            ]));
-        end
-        else if CSpecialText14.Contains(FText) and not IsKeyboardVersion14OrLater then
-        begin
-          ReportError(0, CWARN_TouchLayoutSpecialLabelRequires14,
-            Format('Key "%s" on layout "%s", platform "%s" has a special label "%s" which is only supported in Keyman 14 or later', [
+            Format('Key "%s" on layout "%s", platform "%s" does not have the key type "Special" or "Special (active)" but has the label "%s". This feature is only supported in Keyman 14 or later', [
```

Comment:
Breaking down the condition on a line-by-line basis would probably be clearer.  That conditional is pretty dense, even with the comment at the top, as the bit about "non-special keys in Keyman 13 or earlier" is extremely dense.  (It took me a few minutes to decompress the overall conditional in order to convince myself it was correct as a result.)

- `(CSpecialText10.Contains(FText) or CSpecialText14.Contains(FText))`:  if the key uses a `*special*` text label:
- `not (FKeyType in [tktSpecial, tktSpecialActive])`: keys of type 'special' have always had access to `*special*` labels
    - half of the comment's first line
- `not IsKeyboardVersion14OrLater`: all keys can use `*special*` labels in KMW 14.0+
    - the other half of the comment's first line
- `not CSpecialText14ZWNJ.Contains(FText)`:  if the `*special*` label corresponds to a ZWNJ, it is replaced with a non-`*special*` in versions of KMW before 14.0.

Alternatively, if might be clearer if the `not` lines were reworked a bit.  Instead of `not ____ and not ___`, it would be more easily parsed with a `not (____ or ____)` pattern.  Then, it would effectively read "make a warning on special keys _if_ used outside of these three valid cases".

To state the previous paragraph a bit differently, right now the `if` reads:  
- `if (2 conditions) and (1 condition) and (1 condition) and (1 condition)`
    - There are four top-level groups that seem to have equal weight in this pattern.  
- Alternatively:  `if (2 conditions) and (3 conditions)`
    - There are just two top-level groups, providing a hierarchical grouping that makes parsing the overall conditional easier.
    - The conditions within each group are more easily related to each other than to the conditionals of the other group.

Reply:
I'm not sure that makes the conditional any easier to read, for me: I find negating groups of conditions is harder to map mentally than the current, reasonably flat list of conditions. Think of the first pair of conditions as `[].concat(CSpecialText10, CSpecialText14).includes(foo)` (though that would be much more unwieldy in Delphi code)?

But it's a complex set of conditions, and no matter which way you construct it, it's going to take a bit of mental parsing.